### ISSUE TITLE
File.getContent on HL/Neko/PHP UTF8 problem

### ISSUE THREAD

**Issue by shalmu** [HaxeFoundation/haxe#6930]
File.getContent on HL/Neko/PHP returns a string with trailing BOM in case if the file is encoded in UTF-8 format.

**Comment by shalmu** [comment]
Exactly, has to, on all platforms, but it does it only on "system" platforms and it is different on php and hl. And on js/flash targets there is no BOM. Which, I think, is correct, because of this:
 `JsonParser.parse(File.getContent('a.json'))` should work, right? But it doesn't - json parser immediately throws exception. It works on js/flash platforms though (and should)

Eventually, to make it cross target, I had to write the workaround function:
```
class JaySon{
	public static function parseFile(name:String) {
		function is(n:Int, what:Int) return cont.charCodeAt(n) == what;	
		var cont = File.getContent(name);
		return JsonParser.parse(cont.substr(
			if (is(0, 65279))
				/// looks like a HL target, skipping only first character here:
				1
			else if (is(0, 239) && is(1, 187) && is(2, 191))
				/// it seems to be Neko or PHP, start from position 3:
				3
			else
				/// all other targets, that prepare the UTF string correctly
				0
		));
	}	
}
```
And it took me some time to realize what's going on before I wrote this function.

**Comment by NQNStudios** [comment]
What's the status on this? I just went through quite a long debugging process just to realize my JSON wasn't parsing because the BOM in the file was being included by `getContent()`. Couldn't there at least be an alternate function that strips the BOM on all platforms?

**Comment by ncannasse** [comment]
getContent() should return the BOM, as it's important information and if you do saveContent of getContent it should keep the BOM, not remove it.

Additionnaly, haxe.Json.parse should skip the BOM if it find it as first character.

Both needs to be unit tested.

**Comment by NQNStudios** [comment]
Should I make a separate issue for haxe.json.parse() skipping the BOM? I could also probably make a PR for that if needed.

**Comment by ncannasse** [comment]
@NQNStudios yes, please do also include an unit test so we get reports for all platforms.

**Comment by ncannasse** [comment]
Back to the discussion: we can't make JSON parser BOM compliant because that's not part of the JSON standard. Several possibilities here

choice A: we don't change anything and let users deal with BOM manually

choice B: we add StringTools.withBOM / withoutBOM utilities that adds/remove BOM if-not/if present. The user still has to call them manually to handle it.

choice C: we change getContent to have  ?skipBOM = true, and saveContent to have ?withBOM = false . The asymmetry is voluntary.

**Comment by RealyUniqueName** [comment]
I vote for C. Those methods deal with strings in files, so it's ok to make them BOM-aware.

**Comment by Simn** [comment]
I agree with C.

**Comment by NQNStudios** [comment]
C was my initial instinct, and I'm happy to implement that design.

**Comment by NQNStudios** [comment]
The question remains, which BOMs are important enough to be filtered. In #7643 I tried to catch UTF-8 and UTF-16 and if I understand the error reports correctly, even those two are more complex because the code points end up interpreted differently on different targets. (Some platforms report the UTF-8 BOM as a single 65279, others report it as two separate bytes `[254, 255]` and still others as `[255, 254]`. And the [complete (?) list](https://en.wikipedia.org/wiki/Byte_order_mark#Byte_order_marks_by_encoding) of different BOMs is even more intimidating.

**Comment by ncannasse** [comment]
getContent should take a haxe.io.Encoding as we now do for other Unicode
aware bytes-to-string converters. So far that will only be UTF8 (which is
the default)

**Comment by farteryhr** [comment]
time to revive some kind of text / binary mode switch?
edit: oops there is getBytes. so kind of getRawString and getRepresentedText definition problem.. since BOM is both "a character" and "some header"...
